common/predictive-text/README.md

```diff
@@ -0,0 +1,404 @@
+     */
+    maxLeftContextCodeUnits: 32,
+
+    /**
+     * The maximum amount of code units that the keyboard will provide to
+     * the right of the cursor. The absence of this rule implies 0.
+     * See also, [[supportsRightContexts]].
+     */
+    maxRightContextCodeUnits: 32,
+  }
+}
+```
+
+
+### Message: `ready`
+
+Must be sent from the LMLayer to the keyboard when the LMLayer's model
+as a response to `initialize`. It will send `configuration`, which is
+a plain JavaScript object requesting configuration from the keyboard.
+
+There are only two options defined so far:
+
+```typescript
+interface ReadyMessage {
+  message: 'ready';
+  configuration: {
+    /**
+     * How many UTF-16 code units maximum to send as the context to the
+     * left of the cursor ("left" in the Unicode character stream).
+     *
+     * Affects the `context` property sent in `predict` messages.
+     *
+     * While the left context MUST NOT bisect surrogate pairs, they MAY
+     * bisect graphical clusters.
+     */
+    leftContextCodeUnits: number,
+
+    /**
+     * How many UTF-16 code units maximum to send as the context to the
+     * right of the cursor ("right" in the Unicode character stream).
+     *
+     * Affects the `context` property sent in `predict` messages.
+     *
+     * While the left context MUST NOT bisect surrogate pairs, they MAY
+     * bisect graphical clusters.
+     */
+    rightContextCodeUnits: number,
+  };
+}
+```
+
+
+### Message: `predict`
+
+Sent from the keyboard to the LMLayer whenever a new prediction should
+be generated. This is typically initiated by a key press event. The
+keyboard **SHOULD** track each `predict` message using a [token][Tokens]. The
+token **MUST** be unique across all prediction events. The LMLayer
+**SHOULD** respond to each `predict` message with a `suggestions`
+message. The `suggestions` message **MUST** contain the corresponding
+token as sent in the initial `predict` message.
+
+The keyboard **MUST** send the `context` parameter. The keyboard
+**SHOULD** send the `transform` parameter. The keyboard **MUST** send
+a unique token.
+
+The semantics of the `predict` message **MUST** be from the
+perspective of this sequence of events:
+
+ 1. After the input event is received by the keyboard.
+ 2. Before the keyboard applies the associated `transform` to the buffer.
+
+**NOTE**: The keyboard **MAY** apply the `transform` associated with the
+input event before receiving the corresponding `suggestions` message
+from the LMLayer. The intention is that once the suggestions are displayed,
+the typist may select one of the suggestions in the place of the effects
+of their original input.
+
+**NOTE**: The keyboard **MAY** send the `predict` message after the
+`transform` associated with the input event is applied to the buffer,
+however the semantics **MUST** remain the same—the prediction happens
+from the perspective before the `transform` has been applied.
```

Comment:
To be sure I understand the text here, this does not preclude the possibility of a prediction applying a transform based upon the input event as part of its final transform `Suggestion`, correct?